src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,133 @@
+#include "table_scan.hpp"
+
+#include <memory>
+
+#include "resolve_type.hpp"
+#include "storage/table.hpp"
+
+namespace opossum {
+
+TableScan::TableScan(const std::shared_ptr<const AbstractOperator> in, ColumnID column_id, const ScanType scan_type,
+                     const AllTypeVariant search_value)
+    : AbstractOperator(in), _column_id(column_id), _scan_type(scan_type), _search_value(search_value) {}
+
+TableScan::~TableScan() {}
+
+std::shared_ptr<const Table> TableScan::_on_execute() {
+  const auto& input_table = _input_table_left();
+  const auto& column_type = input_table->column_type(_column_id);
+
+  auto impl = make_unique_by_column_type<BaseTableScanImpl, TableScanImpl>(column_type, input_table, _column_id,
+                                                                           _scan_type, _search_value);
+  return impl->on_execute();
+}
+
+ColumnID TableScan::column_id() const { return _column_id; }
+
+ScanType TableScan::scan_type() const { return _scan_type; }
+
+const AllTypeVariant& TableScan::search_value() const { return _search_value; }
+
+template <typename T>
+TableScan::TableScanImpl<T>::TableScanImpl(std::shared_ptr<const Table> table, ColumnID column_id,
+                                           const ScanType scan_type, const AllTypeVariant search_value)
+    : _table(table), _column_id(column_id), _scan_type(scan_type), _search_value(search_value) {}
+
+template <typename T>
+void TableScan::TableScanImpl<T>::_scanColumn(std::shared_ptr<PosList> position_list,
+                                              std::shared_ptr<ValueColumn<T>> vc, const ChunkID chunk_id) {
+  const auto& values = vc->values();
+  const auto compare_value = opossum::type_cast<T>(_search_value);
+
+  const auto accessor = [&](size_t chunk_offset) { return values[chunk_offset]; };
+  _appendPositionList(position_list, accessor, values.size(), compare_value, chunk_id, false);
+}
+
+template <typename T>
+void TableScan::TableScanImpl<T>::_scanColumn(std::shared_ptr<PosList> position_list,
+                                              std::shared_ptr<DictionaryColumn<T>> dc, const ChunkID chunk_id) {
+  const auto& id_values = dc->attribute_vector();
+  ValueID dict_id = INVALID_VALUE_ID;
+  bool all_false = false;
+  bool all_true = false;
+  const auto compare_value = opossum::type_cast<T>(_search_value);
+
+  switch (_scan_type) {
```

Comment:
Please consider adding comments to your code to help others - especially if they have not implemented the same exercise - understand your intentions. We noticed that you did not comment your code. Especially here this would be nice to help the reader understand why you are doing this.